### ISSUE TITLE
Need guidance on enabling/disabling submit buttons

### ISSUE THREAD

**Issue by mcarrano** [patternfly/patternfly-design#388]
#### 1. Choose the type of contribution for this pattern:
- [ ] This is a new pattern 
- [x ] This is an enhancement on an existing pattern
Existing Errors and Validations page should include guidance on when to perform validation and disable submit buttons.  Some suggestions include:

    * when one or more required fields are omitted, we recommend disabling the action button
    * when possible, it's best to perform inline validation as the widget loses focus so that the user is immediately informed of invalid information
   *  if there are any invalid entries on the page,we recommend disabling the action button
    * if additional input validation is performed once the action button is clicked, we recommend staying in context while showing input validation errors, so the user has the opportunity to view and modify the original form

Related PF Story: https://patternfly.atlassian.net/browse/PTNFLY-2355

**Comment by yih-wang** [comment]
@mcarrano  @LHinson I did some research on improving recommendations for form validation. Please have a look and welcome any feedback.
https://docs.google.com/a/redhat.com/document/d/1gXWWj42h_yCPDh8_Xn6ZWhc2AyFSjV6tfHojg7SfLo4/edit?usp=sharing

**Comment by mcarrano** [comment]
@LHinson I am comfortable with these recommendations. Please take a look and let me know if you agree. I think that next steps would be to add these recommendations to Errors and Validation pattern (http://www.patternfly.org/pattern-library/forms-and-controls/errors-and-validation/) after we run by UXS team for preliminary review.

**Comment by mcarrano** [comment]
The recommendations outlined by @yih-wang here (https://docs.google.com/a/redhat.com/document/d/1gXWWj42h_yCPDh8_Xn6ZWhc2AyFSjV6tfHojg7SfLo4/edit?usp=sharing) do conflict with earlier recommendations as stated on the Buttons on Forms page here: https://docs.google.com/a/redhat.com/document/d/1gXWWj42h_yCPDh8_Xn6ZWhc2AyFSjV6tfHojg7SfLo4/edit?usp=sharing

@serenamarie125 @catrobson @maryclarke @jeperry can you please take a look at these documents and let us know what you think about this issue? How are we implementing form validation in products you have worked on? The issue seems to be whether a submit button should be disabled until all conditions on a form are met or whether we allow the user to submit and then correct any resulting errors. I suspect that the best approach might be contingent on when and how data validation is performed. It may just be that we need to spell out better what the various scenarios are and how to handle rather than give a general statement.

@lizsurette , you may also want to take a look as I believe you wrote the Buttons on Forms guidelines.

**Comment by catrobson** [comment]
@mcarrano Thanks for pointing us to this.  I think the disable of the submit button came from teams who were working on UIs like the RHCI installer where there was a one-page experience and the button and content were always visible.  In these situations, the disabling of the button does make sense.  I agree with @yih-wang's suggestion that in the cases of long scrolling-page forms, this does not make sense.  I'd like to see us provide guidance for both situations in PatternFly in our guidelines - it might be that wizards get special treatment.

**Comment by lizsurette** [comment]
@mcarrano Some additional comments to @catrobson's...

1) "When a required field is not filled or a wrong format is input, we recommend showing a validation as the widget loses focus." - I think this is a good idea, but one thing to be careful about is that all of the widgets support something like this consistently.

2) I have a case similar to the RHCI UI that Cat mentions for OpenStack where we have somewhat of a 1 page wizard and we want to make it clear that the user needs to complete a few of the steps before "Deploying". We disable the "Deploy" button until the previous necessary steps are completed and then it becomes enabled. So I do think having a specific situation where maybe it's a wizard use case that supports disabling the button would be good. See this image for the example:
https://lizsurette.github.io/OpenStack-Design/tripleo-ui/1-deployment-workflow-for-edge-technician/img/2017-7-24-TripleO-UI_13_Jacques11.png

3) Heads up @mcarrano, I think the second link in your comment should be this:
http://www.patternfly.org/pattern-library/forms-and-controls/buttons-on-forms/
With respect to this, I think we should absolutely update these guidelines to follow whatever we decide for this GitHub issue. I have no attachment to the current guidelines stated in the "Enabling and Disabling Actions" section :)

Hope this helps!
Liz

**Comment by yih-wang** [comment]
@mcarrano Agree with @catrobson and @lizsurette's comments. We could have a general rule for common forms and a specific rule for wizards. I added a paragraph to the end of the doc (https://docs.google.com/a/redhat.com/document/d/1gXWWj42h_yCPDh8_Xn6ZWhc2AyFSjV6tfHojg7SfLo4/edit?usp=sharing). If we all agree the solution, I can move forward with updating the pattern in PatternFly.

**Comment by mcarrano** [comment]
@yih-wang Sorry to take so long to get back to you on this, but yes, let's move forward with your suggestion and update the pattern accordingly by sending a PR.

**Comment by yih-wang** [comment]
@mcarrano @LHinson Added the new rules to the existing [Errors and Validations](http://www.patternfly.org/pattern-library/forms-and-controls/errors-and-validation/) page. A PR was sent: #465 . Plz have a review, thanks.

**Comment by mcarrano** [comment]
@yih-wang  Please see http://www.patternfly.org/pattern-library/forms-and-controls/buttons-on-forms/ As part of this PR or as a separate PR, so you think we need to update this pattern at all? See the section on Enabling/Disabling Actions.

**Comment by yih-wang** [comment]
@mcarrano Yes, I think we should clarify it will happen only in one-page design. The button should always be enabled in other cases. I'll update it together in #465 .

**Comment by mcarrano** [comment]
Merged this PR. Validate changes on site before closing.

**Comment by mcarrano** [comment]
Verified.
